### PR TITLE
api-server: initial implementation of basic features

### PR DESCRIPTION
**Pull request by koonpeng**
## What's new

This lays the groundwork for many of the features required in the backend:

* subscribing to rmf events
* writing events/logs to a sql database
* restoring state from database
* socketio endpoint
* rest endpoint
* ci
  * tests
  * coverage
  * linting
  * formatting

Currently, this poc only supports door states and building map. It is hoped that with the groundwork done, it will be much easier to add all the other topics as needed.

TODO: dockerfile

More detail copied from the docs:

---

# Architecture

The simplest design is a single app that does all the needed functions, At first this seems fine until we start thinking about scaling it, the simplest approach would be to try and run multiple instances of the app, each subscribing to RMF topics and pushing them to clients. However, there are multiple problems if we simply try to replicate the app

* RMF does not provide all the data that we need. Some events have to be derived based on time-series data and/or multiple events.
* Each instance listens to RMF event without synchronization with others, if one instance miss or receive some messages late, it may cause the derived events to be different.
* If there ever is different history of events between instances, each will write to run their version into the database, we would get database conflicts there are not easily resolved.
* Each instance will be hitting the database constantly, particularly they will be hitting it with many duplicated write operations, traditional SQL-based databases does not work well in such workloads, eventual consistency based database will scale better but may create data conflicts as mentioned above.

In order to tackle the above mentioned issues, this is the current proposed architecture:

dashboard-data-backend-proposal

* A single instance socketio based app will subscribe to RMF events and write them to a database.
* All events published will be "stickied" or "transient local" in ros terms, it means any client that subscribes to the events will immediately receive the latest state.
* Additional socketio apps will subscribe to this main app and act as proxy/fanout.
* Additional rest based apps can connect to the proxies or directly to the main app.
* These additional apps will not write RMF events data to the database, however they may write non RMF data (e.g. user preferences).
* These additional apps will not derive any events, all derived events shall be received from the main socketio app.
* Dataflow from the other direction (from client to RMF) will all have to go through the main socketio app. However, it is assumed that RMF will be the bottleneck anyway.

### Failover/redundancy
The main socketio app will be a single point of failure.
* This is somewhat mitigated as it is able to use the database to restore it's state.
* Technologies like kubernetes can automatically restart a container when it crashes, minimizing downtime.
* Technically it is possible to run failover instances of the main app, we just need _very_ complicated logic to synchronize states across them.

---

## Self-checks

- [ ] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

### CODE REVIEW

File: packages/api-server/README.md [hunk 1]
```diff
@@ -1,19 +1,72 @@
...
-Let's create a virtual environment named `venv` for all this Python magic:
+Install pipenv
+
+```bash
+pip3 install pipenv
 ```
-sudo apt install python3-venv
-python3 -m venv --system-site-packages .venv
+
+If not already done so, [bootrap](../../README.md#bootstrap) the project, you can use
```
Comment (codebot) [inline packages/api-server/README.md:9]: trivial typo: `bootrap` -> `bootstrap`

File: packages/api-server/rest_server/app.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+    app.include_router(
+        building_map_router(sio, logger.getChild("building_map")),
+        prefix="/building_map",
+    )
+
+    logger.info("started app")
+
+
+# There is a big with uvicorn and socketio that causes
```
Comment (matiasbavera) [inline packages/api-server/rest_server/app.py:41]: big -> bug . I think your solution is super valid.

**Review by matiasbavera — COMMENTED** [review]
Awesome! Thanks for this, left a couple of comments =)

File: packages/api-server/api_server/repositories/sql.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    async def write_door_state(self, rmf_door_state: RmfDoorState):
+        door_state = DoorState.from_rmf(rmf_door_state)
+        await door_state.save()
+        self.logger.debug(f'written door_state ("{door_state.door_name}") to database')
+
+    async def read_door_states(self) -> Sequence[RmfDoorState]:
+        # FIXME: This is probably inefficient but I don't know how to write a good filter.
+        latest_states = []
+        doors = await DoorState.all().distinct().values("door_name")
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/repositories/sql.py:25]: I tried it but it doesn't work, what we need is roughly "get latest door state for each distinct door", `DoorState.group_by("door_name").values("door_name", "door_time")` returns distinct door name but unsorted. `DoorState.order_by("-door_time").group_by("door_name").values("door_name", "door_time")` performs `group_by` first before `order_by`, so it doesn't return the latest state.

Apparently what we need is either MIN(), MAX(), DISTINCT ON, PARTITION OVER or other engine specific queries.

File: packages/api-server/api_server/repositories/sql.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    def __init__(self, logger: logging.Logger = None):
+        self.logger = logger or logging.getLogger(self.__class__.__name__)
+
+    async def write_door_state(self, rmf_door_state: RmfDoorState):
+        door_state = DoorState.from_rmf(rmf_door_state)
+        await door_state.save()
+        self.logger.debug(f'written door_state ("{door_state.door_name}") to database')
+
+    async def read_door_states(self) -> Sequence[RmfDoorState]:
+        # FIXME: This is probably inefficient but I don't know how to write a good filter.
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/repositories/sql.py:23]: That is only if we are interested in the FK right? Suppose we have a FK which points to the `Door` record, but we are not really interested in the door location etc so we shouldn't prefetch the FK?

File: packages/api-server/api_server/models/door_state.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+from rmf_door_msgs.msg import DoorState as RmfDoorState
+from tortoise import fields
+from tortoise.models import Model
+
+from ..ros_time import py_to_ros_time, ros_to_py_datetime
+from .door_mode import DoorModeEnum
+
+
+class DoorState(Model):
+    # TODO: foreign key with Door
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/models/door_state.py:11]: Mostly yea, but that is mostly an experimental, I didn't end up using the mixins because turns out every single field in `door_states` needs some model relation, it remains to be seen how useful the mixins will be for other models. Maybe it is possible to generate the relations base on the message but I don't know how to do it.

File: packages/api-server/api_server/repositories/sql.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import logging
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/repositories/sql.py:1]: I'm not very certain on this, for one, the model themselves already expose the query functions, I guess we still need this to translate between ros messages <-> sql models, and I guess it's also nice to consolidate the queries here.

`sql.py` is just short for `sql_repository.py` because I feel that it is redundant to repeat the repo
`from .repositories.sql_repository import SqlRepository`
vs
`from .repositories.sql import SqlRepository`

File: package.json [hunk 7]
```diff
@@ -2,6 +2,7 @@
   "name": "root",
   "private": true,
   "scripts": {
+    "install": "python3 -m pipenv install -d",
     "bootstrap": "node scripts/bootstrap.js"
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply package.json:6]: thats weird, somehow `--site-p` is truncated?

File: packages/api-server/rest_server/app.py (hunk 2 above)
Comment (koonpeng) [reply packages/api-server/rest_server/app.py]: fyi, the "solution" I did is to simply split them up into 2 ASGI app, because of the bug, we can't run fastapi and socketio together (we could, but need handle cleanup of both apps manually). So the easier solution is to simply just split them up, its also ends up being more "micro-service".

Comment (codebot) [reply packages/api-server/rest_server/app.py]: everybody loves micro-services these days

File: packages/api-server/api_server/repositories/sql.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/api-server/api_server/repositories/sql.py:23]: > That is only if we are interested in the FK right?

Yeap, and 

> Suppose we have a FK which points to the Door record, but we are not really interested in the door location etc so we shouldn't prefetch the FK?

Yeah, you're right

File: packages/api-server/api_server/models/door_state.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/api-server/api_server/models/door_state.py:11]: Got it. 

> Maybe it is possible to generate the relations base on the message but I don't know how to do it.

You mean something like can create models and relations by introspecting existing messages (there is a tool for that but it only works with SQL tables =/ ). Just thinking out loud here, if we do that we also need to make a new "migration" every time a msg format changes, before receiving the message.

File: packages/api-server/api_server/repositories/sql.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/api-server/api_server/repositories/sql.py:1]: Oh, ok, that makes sense. Thanks

File: packages/api-server/api_server/repositories/sql.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply packages/api-server/api_server/repositories/sql.py:25]: Don't worry I'll give it a look when I make it work on my local env =).

Also, it's ok because those queries are supposed to be lazy loading

File: package.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (matiasbavera) [reply package.json:6]: Yeah, that's right, I tried with `python3 -m pipenv install -d --site-packages` and `python3 -m pipenv install -d --site-packages` and neither are working. `python3 -m pipenv install -d` it's working but I think it's not installing all the packages that we want. 

`pipenv, version 11.9.0`
`Python 3.8.5`
`Ubuntu 20`

File: packages/api-server/package.json [hunk 8]
```diff
@@ -0,0 +1,14 @@
+{
+  "name": "api-server",
+  "version": "0.0.0",
+  "description": "dummy package",
+  "private": true,
+  "scripts": {
+    "install": "python3 -m pipenv install -d --site-packages",
```
Comment (matiasbavera) [inline packages/api-server/package.json:7]: Having an error on install, described above

File: packages/api-server/api_server/rmf_io/transport.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        self.door_states_sub = self.ros2_node.create_subscription(
+            DoorState, "door_states", self.rmf_gateway.door_states.on_next, 10
+        )
+
+        self.building_map_sub = self.ros2_node.create_subscription(
+            BuildingMap,
+            "map",
+            self.rmf_gateway.building_map.on_next,
+            rclpy.qos.QoSProfile(
+                history=rclpy.qos.HistoryPolicy.KEEP_LAST,
```
Comment (matiasbavera) [inline packages/api-server/api_server/rmf_io/transport.py:29]: Dump question, Should we use the `KEEP_LAST` since we already have a database that saves the states? Or it's better to have it both?

**Review by matiasbavera — COMMENTED** [review]
Thank KP, the code looks good to me. I'm having problems installing and make it work explained above.

File: packages/api-server/api_server/rmf_io/transport.py (hunk 9 above)
Comment (koonpeng) [reply packages/api-server/api_server/rmf_io/transport.py:29]: rmf only publishes the building map once, if we don't use `KEEP_LAST`, we must start the server before rmf else we will never receive any building map.

Comment (matiasbavera) [reply packages/api-server/api_server/rmf_io/transport.py:29]: Cool, thanks for the feedback.

**Review by codebot — APPROVED** [review]

File: packages/api-server/package.json (hunk 8 above)
Comment (koonpeng) [reply packages/api-server/package.json:7]: I'm not able to reproduce this, what are your versions? mine is 
python3: 3.8.5
pipenv: 2020.11.15
pip3: 20.0.2

File: packages/api-server/api_server/models/door_state.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/api-server/api_server/models/door_state.py:11]: > there is a tool for that but it only works with SQL tables =/ 

Thats very interesting, perhaps we can find a way to make it generate tortoise-orm models as well?

> Just thinking out loud here, if we do that we also need to make a new "migration" every time a msg format changes, before receiving the message.

I think this also brings up a point in regards to updating message, the impact and migration strategy should be carefully analyzed in their respective PR, in some cases, it might not be possible for `rmf-web` to perform the migration. For example, if the new messages contains a *required* field, we can't migrate by creating a new nullable column, that would violate the schema, so the message PR would need to describe how to fill in the new fields for existing records, or simply acknowledge that updating to the new version will be a breaking change, not just in terms of API/ABI, but also all database will be invalidated and need to restart from an empty state.

File: README.md [hunk 10]
```diff
@@ -29,6 +29,11 @@ curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.37.2/install.sh | bash
 nvm install 12
 ```
 
+A recent version of pipenv is needed, the system packaged version is too old.
+```bash
+pip3 install pipenv
```
Comment (matiasbavera) [inline README.md:34]: Thanks man, this worked for me `pip install --upgrade pipenv`

**Review by matiasbavera — APPROVED** [review]
LGTM! Thanks for this KP! Working correctly for me too! Awesome work.
